includes/admin/upgrades/v3/class-data-migrator.php

```diff
@@ -1306,6 +1304,40 @@ public static function orders( $data = null ) {
 		do_action( 'edd_30_migrate_order', $order_id, $payment_meta, $meta );
 	}
 
+	/**
+	 * Retrieves a valid price ID for a given cart item.
+	 * If the product does not have variable prices, then `null` is always returned.
+	 * If the supplied price ID does not match a price ID that actually exists, then the default
+	 * variable price is returned instead of the supplied one.
+	 *
+	 * @since 3.0
+	 *
+	 * @param array $cart_item Array of cart item details.
+	 *
+	 * @return int|null
+	 */
+	protected static function get_valid_price_id_for_cart_item( $cart_item ) {
+		// If the product doesn't have variable prices, just return `null`.
+		if ( ! edd_has_variable_prices( $cart_item['id'] ) ) {
+			return null;
+		}
+
+		// Get the price ID that's set to the cart item right now.
+		$price_id = isset( $cart_item['item_number']['options']['price_id'] ) && is_numeric( $cart_item['item_number']['options']['price_id'] )
+			? absint( $cart_item['item_number']['options']['price_id'] )
+			: null;
+
+		// Now let's confirm it's actually a valid price ID.
+		$variable_prices = edd_get_variable_prices( $cart_item['id'] );
+		if ( ! is_array( $variable_prices ) || empty( $variable_prices ) ) {
```

Comment:
Tested on a migration of my own site data and it worked beautifully. Going to test EDD as well, but what if this early return was moved above the `$price_id` definition since that variable isn't actually used until the very end?

Reply:
Looks like this is working nicely on EDD data as well, even with retired products, which do have a `0` price ID involved. I can approve with or without moving the early return but will wait to see what you think.

Reply:
Good idea. :+1: https://github.com/awesomemotive/easy-digital-downloads/pull/8927/commits/3b190bbf4f17e0d83333236387bdaa088c83ceef